docs/life-of-a-request.md

```diff
@@ -0,0 +1,200 @@
+  useful to check a link without downloading a potentially large link
+* OPTIONS (handled automatically, see more below)
+* POST for everything else
+
+Of these, PUT, DELETE, HEAD, OPTIONS, and GET are *idempotent*, which
+means that you can send the request multiple times and get the same
+state on the server. POST is not idempotent--if I send a message
+multiple times, I will get multiple message. PATCH is special--it can be
+idempotent, and we like it if you write API endpoints where it is
+idempotent, as much as possible.
+
+### PUT vs POST vs PATCH
+
+The difference between these verbs is subtle at first glance, but
+important.
+
+Let's say you have something (the HTTP RFC calls them "resources") like
+a realm emoji. An emoji has two attributes: a name and a url for the
+emoji image. Let's pretend there's also an emoji alternative name
+attribute.
+
+In this case, we should avoid POST, since we have a unique identifier:
+the name. We don't want to create multiple emoji with the same name--
+otherwise it would be ambiguous which emoji we want to select when
+composing a message.
+
+If we want create the emoji, we should use PUT, since we're replacing
+the entire resource at the unique ID (name). If we wanted to simply
+update some attribute of the emoji, either the image url or the
+alternative name, we should use PATCH. Otherwise, to update the image
+url without clearing the alternative name would require knowing the
+alternative name of the emoji we want to upate. If we wanted to make it
+so that you couldn't update the image url without changing the
+alternative name, or at least knowing it, we could enforce using PUT
+for updates to the realm emoji.
+
+http://restful-api-design.readthedocs.io/en/latest/methods.html can be
+a good read.
+
+### OPTIONS
+
+This HTTP verb will yield the allowed methods 
+OPTIONS https://zulip.tabbott.net/api/v1/users
+
+will yield, as a HTTP header:
+
+`Allow: POST, GET`
+
+You can see this reflected in [zproject/urls.py](https://github.com/zulip/zulip/blob/master/zproject/urls.py):
+
+    url(r'^users$', 'zerver.lib.rest.rest_dispatch',
+        {'GET': 'zerver.views.users.get_members_backend',
+         'POST': 'zerver.views.users.create_user_backend'}),
+
+In this way, the API is partially self-documenting.
+
+### Legacy endpoints are used by the web client
+
+Many of endpoints from our legacy JSON API are written without REST in
+mind. They are still largely used by the web client, and use POST.
+
+You can see them in [zproject/legacy_urls.py](https://github.com/zulip/zulip/blob/master/zproject/legacy_urls.py).
+
+### Webhook integrations may not be RESTful
+
+Webhook--Zulip endpoints that are called by other services, especially
+for integrations, have to conform to the service's request format. Thus,
+they may not use REST, and are likely to use either POST or GET without
+the REST discipline of idempotency.
+
+TODO: mention webhook decorator
```

Comment:
Yep, api_key_only_webhook_view.


        -Tim Abbott

On Sun, Jul 17, 2016 at 1:29 PM, acrefoot <notifications@github.com> wrote:

> In docs/life-of-a-request.md
> <https://github.com/zulip/zulip/pull/1302#discussion_r71088833>:
>
> > +
> > +### Legacy endpoints are used by the web client
> > +
> > +Many of endpoints from our legacy JSON API are written without REST in
> > +mind. They are still largely used by the web client, and use POST.
> > +
> > +You can see them in [zproject/legacy_urls.py](https://github.com/zulip/zulip/blob/master/zproject/legacy_urls.py).
> > +
> > +### Webhook integrations may not be RESTful
> > +
> > +Webhook--Zulip endpoints that are called by other services, especially
> > +for integrations, have to conform to the service's request format. Thus,
> > +they may not use REST, and are likely to use either POST or GET without
> > +the REST discipline of idempotency.
> > +
> > +TODO: mention webhook decorator
>
> @timabbott <https://github.com/timabbott> I'm not sure which webhook
> decorator you were talking about.
>
> Do you mean @api_key_only_webhook_view, or the webhook-specific parts of
> validate_api_key?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/1302/files/9281eb81f79b0753a8df4fb2349910a9ab62ead2..2bbaf5a43e58311c8a922aed3770ab2bb630c02c#r71088833>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACnm2hir-YgFvFTB9qR-KV-qx_ISndVcks5qWpCMgaJpZM4JN49d>
> .
>